pkg/util/deployment/deployment.go

```diff
@@ -444,3 +444,35 @@ func WaitForObservedDeployment(getDeploymentFunc func() (*extensions.Deployment,
 		return deployment.Status.ObservedGeneration >= desiredGeneration, nil
 	})
 }
+
+// ResolveFenceposts resolves both maxSurge and maxUnavailable. This needs to happen in one
+// step. For example:
+//
+// 2 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1), then old(-1), then new(+1)
+// 1 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1)
+// 2 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
+// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
+func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
+	surge, err := intstrutil.GetValueFromIntOrPercent(maxSurge, desired, true)
+	if err != nil {
+		return 0, 0, err
+	}
+	if surge == 0 {
+		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
+		// then we want to allow scaling down and not block rollouts.
```

Comment:
Agree on not rounding up in general on maxUnavailable.  We just don't
want to go to available 0 ever when maxUnavailable != 100%

For some context here:

The scenario (which I expect is not uncommon) is our ingress L7
routers.  They are pods, running on host network, spread across a set
of nodes.  Those nodes have a keepalived and VVRP pool of virtual IPs
that get moved across nodes.  We can't surge (because there are only
so many hosts that qualify), but we don't want to set a numeric
maxUnavailable (because we only want to drop a few at a time, not all
of them).  We chose 25% because we expect in general that admins would
run sets of 3-4 routers in heavy load cases, and 2 routers otherwise.

In the future, we want to switch to daemon set for these, but would
need all the same guarantees.